src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,381 @@
+        new MeterListener(
+            InstrumentPublished =
+                fun i l ->
+                    if i.Meter = meter && i.Description = instanceId then
+                        l.EnableMeasurementEvents(i)
+        )
+
+    do
+        listener.SetMeasurementEventCallback<int64>(fun k v _ _ -> Interlocked.Add(current.GetOrAdd(k, ref 0L), v) |> ignore)
+        listener.Start()
+
+    member val CacheId = instanceId
+
+    member val RecentStats = "-" with get, set
+
+    member this.TryUpdateStats(clearCounts) =
+        let stats =
+            try
+                let ratio =
+                    float current[hits].Value / float (current[hits].Value + current[misses].Value)
+                    * 100.0
+
+                [
+                    for i in current.Keys do
+                        let v = current[i].Value
+
+                        if v > 0 then
+                            $"{i.Name}: {v}"
+                ]
+                |> String.concat ", "
+                |> sprintf "%s | hit ratio: %s %s" this.CacheId (if Double.IsNaN(ratio) then "-" else $"%.1f{ratio}%%")
+            with _ ->
+                "!"
+
+        if clearCounts then
+            for r in current.Values do
+                Interlocked.Exchange(r, 0L) |> ignore
+
+        if stats <> this.RecentStats then
+            this.RecentStats <- stats
+            true
+        else
+            false
+
+    member this.Dispose() = listener.Dispose()
+
+    static member GetStats(cache: ICacheEvents) =
+        instrumentedCaches[cache].TryUpdateStats(false) |> ignore
+        instrumentedCaches[cache].RecentStats
+
+    static member GetStatsUpdateForAllCaches(clearCounts) =
+        [
+            for i in instrumentedCaches.Values do
+                if i.TryUpdateStats(clearCounts) then
+                    i.RecentStats
+        ]
+        |> String.concat "\n"
+
+    static member AddInstrumentation(cache: ICacheEvents) =
+        instrumentedCaches[cache] <- new CacheInstrumentation(cache)
+
+    static member RemoveInstrumentation(cache: ICacheEvents) =
+        instrumentedCaches[cache].Dispose()
+        instrumentedCaches.TryRemove(cache) |> ignore
+
+module Cache =
+
+    // During testing a lot of compilations are started in app domains and subprocesses.
+    // This is a reliable way to pass the override to all of them.
+    [<Literal>]
+    let private overrideVariable = "FSHARP_CACHE_OVERRIDE"
+
+    /// Use for testing purposes to reduce memory consumption in testhost and its subprocesses.
+    let OverrideMaxCapacityForTesting () =
+        Environment.SetEnvironmentVariable(overrideVariable, "true", EnvironmentVariableTarget.Process)
+
+    let Create<'Key, 'Value when 'Key: not null and 'Key: equality> (options: CacheOptions) =
+
+        let options =
+            match Environment.GetEnvironmentVariable(overrideVariable) with
+            | null -> options
+            | _ -> { options with MaximumCapacity = 1024 }
```

Comment:
So we have got:

CacheOptions.Default.MaximumCapacity => That one is **never** used ... ?

createTypeSubsumptionCache depending on the options.

and this piece overrides it agnostic of the options (I guess it assumes that CI run is always one-off and isolated, and works with a very small cache intentionally).


?

Reply:
Yes the idea is to override only the MaximumCapacity when run in testhost process. We call `OverrideMaxCapacityForTesting ()` at the start of xUnit test run. The env var should propagate to app domains on net472, and even subprocesses that we start a lot.

Reply:
Ok that makes sense.
I agree to rather have small cache than no-cache at all for CI, so that we aspise to keep a higher level of dogfooding